File: heap-web/src/test/java/com/github/ol_loginov/heaplibweb/boot_test/DatabaseTest.java
package com.github.ol_loginov.heaplibweb.boot_test;

import com.github.ol_loginov.heaplibweb.boot.RepositoryConfig;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.boot.autoconfigure.context.ConfigurationPropertiesAutoConfiguration;
import org.springframework.boot.autoconfigure.flyway.FlywayAutoConfiguration;
import org.springframework.boot.autoconfigure.jdbc.DataSourceAutoConfiguration;
import org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaAutoConfiguration;
import org.springframework.boot.autoconfigure.transaction.TransactionAutoConfiguration;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.context.annotation.Import;
import org.springframework.test.annotation.Rollback;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.transaction.PlatformTransactionManager;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.support.TransactionTemplate;

@ExtendWith(MockitoExtension.class)
@ActiveProfiles({"local", "test"})
@SpringBootTest(classes = {ConfigurationPropertiesAutoConfiguration.class, DatabaseTest.DatabaseTestContext.class})
@Transactional
@Rollback
public abstract class DatabaseTest {
	@Configuration
	@Import({TransactionAutoConfiguration.class, DataSourceAutoConfiguration.class, FlywayAutoConfiguration.class, HibernateJpaAutoConfiguration.class, RepositoryConfig.class})
	static class DatabaseTestContext {
		@Bean
		public TransactionTemplate transactionTemplate(PlatformTransactionManager transactionManager) {
			return new TransactionTemplate(transactionManager);
		}
	}

	@PersistenceContext
	protected EntityManager entityManager;
}

File: heap-web/src/test/java/com/github/ol_loginov/heaplibweb/boot_test/DatabaseTest.kt
package com.github.ol_loginov.heaplibweb.boot_test

import com.github.ol_loginov.heaplibweb.boot.RepositoryConfig
import com.github.ol_loginov.heaplibweb.repository.HeapFile
import com.github.ol_loginov.heaplibweb.repository.HeapFileRepository
import jakarta.inject.Inject
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.boot.autoconfigure.EnableAutoConfiguration
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.context.annotation.Configuration
import org.springframework.context.annotation.Import
import org.springframework.jdbc.core.simple.JdbcClient
import org.springframework.test.annotation.Rollback
import org.springframework.test.context.ActiveProfiles
import org.springframework.transaction.annotation.Transactional
import org.springframework.transaction.support.TransactionOperations

@ExtendWith(MockitoExtension::class)
@EnableAutoConfiguration
@ActiveProfiles("local", "test", "test-local")
@SpringBootTest(classes = [DatabaseTest.DatabaseTestContext::class])
@Transactional
@Rollback
abstract class DatabaseTest {
    @Configuration
    @Import(RepositoryConfig::class)
    class DatabaseTestContext {
//        @Bean
//        fun transactionTemplate(transactionManager: PlatformTransactionManager?): TransactionTemplate {
//            return TransactionTemplate(transactionManager!!)
//        }
    }

    @Inject
    protected lateinit var jdbc: JdbcClient
    @Inject
    protected lateinit var heapFileRepository: HeapFileRepository
    @Inject
    protected lateinit var transactionOperations: TransactionOperations

    private val heapsToDrop = mutableListOf<Int>()

    protected fun dropScopeAfterTest(heap: HeapFile?) {
        heap?.let { heapsToDrop.add(it.id) }
    }

    protected fun dropScopeAfterTest(heapId: Int) {
        heapsToDrop.add(heapId)
    }

    @AfterEach
    fun clearTest() {
        heapsToDrop.forEach { heap ->
            transactionOperations.executeWithoutResult {
                heapFileRepository.findById(heap)?.let { entity ->
                    heapFileRepository.getHeapRepositories(entity).dropTables()
                }
            }
        }
    }
}